benchmark/bls.cpp

```diff
@@ -0,0 +1,417 @@
+                     ("actor", name("payloadless"_n))
+                     ("permission", name(config::active_name))
+               }))
+               ("data", fc::mutable_variant_object()
+               )
+            })
+        );
+      trx = std::make_unique<signed_transaction>();
+      abi_serializer::from_variant(pretty_trx, *trx, chain->get_resolver(), abi_serializer::create_yield_function( chain->abi_serializer_max_time ));
+      chain->set_transaction_headers(*trx);
+      trx->sign( chain->get_private_key( "payloadless"_n, "active" ), control->get_chain_id() );
+
+      // construct a packed transaction
+      ptrx = std::make_unique<packed_transaction>(*trx, eosio::chain::packed_transaction::compression_type::zlib);
+
+      // build transaction context from the packed transaction
+      timer = std::make_unique<platform_timer>();
+      trx_timer = std::make_unique<transaction_checktime_timer>(*timer);
+      trx_ctx = std::make_unique<transaction_context>(*control, *ptrx, ptrx->id(), std::move(*trx_timer));
+      trx_ctx->max_transaction_time_subjective = fc::microseconds::maximum();
+      trx_ctx->init_for_input_trx( ptrx->get_unprunable_size(), ptrx->get_prunable_size() );
+      trx_ctx->exec();
+
+      // build apply context from the control and transaction context
+      apply_ctx = std::make_unique<apply_context>(*control, *trx_ctx, 1);
+
+      // finally build the interface
+      interface = std::make_unique<webassembly::interface>(*apply_ctx);
+   }
+
+   std::unique_ptr<tester>                      chain;
+   controller*                                  control;
+   std::unique_ptr<signed_transaction>          trx;
+   std::unique_ptr<packed_transaction>          ptrx;
+   std::unique_ptr<platform_timer>              timer;
+   std::unique_ptr<transaction_checktime_timer> trx_timer;
+   std::unique_ptr<transaction_context>         trx_ctx;
+   std::unique_ptr<apply_context>               apply_ctx;
+   std::unique_ptr<webassembly::interface>      interface;
+};
+} // namespace eosio::chain
+
+
+namespace benchmark {
+
+// utilility to create a random scalar
+std::array<uint64_t, 4> random_scalar()
+{
+   std::random_device rd;
+   std::mt19937_64 gen(rd());
+   std::uniform_int_distribution<uint64_t> dis;
+
+   return {
+      dis(gen) % bls12_381::fp::Q[0],
+      dis(gen) % bls12_381::fp::Q[1],
+      dis(gen) % bls12_381::fp::Q[2],
+      dis(gen) % bls12_381::fp::Q[3]
+   };
+}
+
+// some functions like pairing/exp are expensive. set a hard limit to
+// the number of runs such that the transanction does not reach deadline
+constexpr uint32_t num_runs_limit = 200;
+
+// utilility to create a random g1
+bls12_381::g1 random_g1()
+{
+   std::array<uint64_t, 4> k = random_scalar();
+   return bls12_381::g1::one().mulScalar(k);
+}
+
+// utilility to create a random g2
+bls12_381::g2 random_g2()
+{
+   std::array<uint64_t, 4> k = random_scalar();
+   return bls12_381::g2::one().mulScalar(k);
```

Comment:
Does using `one()` provide a good distribution here? Would it be better to have random for both? Same question for g1.

Reply:
I copied this code from the BLS library's unittest.cpp. As `k` is randomized, it might be random enough for benchmark/tests.